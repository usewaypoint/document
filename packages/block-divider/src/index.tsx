import React, { CSSProperties } from 'react';
import { z } from 'zod';

function zColor() {
  return z.string().regex(/^#[0-9a-fA-F]{6}$/);
}

export const DividerPropsSchema = z.object({
  style: z
    .object({
      backgroundColor: zColor().optional().nullable(),
      padding: z
        .object({
          top: z.number(),
          bottom: z.number(),
          right: z.number(),
          left: z.number(),
        })
        .optional()
        .nullable(),
    })
    .optional(),
  props: z
    .object({
      lineColor: zColor().optional().nullable(),
      lineHeight: z.number().optional().nullable(),
    })
    .optional(),
});

export type DividerProps = z.infer<typeof DividerPropsSchema>;

export function Divider({ style, props }: DividerProps) {
  const st: CSSProperties = {
    padding: getPadding(style),
    backgroundColor: style?.backgroundColor ?? undefined,
  };
  const borderTopWidth = props?.lineHeight ?? 1;
  const borderTopColor = props?.lineColor ?? '#333333';
  return (
    <div style={st}>
      <hr
        style={{
          width: '100%',
          border: 'none',
          borderTop: `${borderTopWidth}px solid ${borderTopColor}`,
          margin: 0,
        }}
      />
    </div>
  );
}

function getPadding(style: DividerProps['style']) {
  const value = style?.padding;
  if (!value) {
    return undefined;
  }
  return `${value.top}px ${value.right}px ${value.bottom}px ${value.left}px`;
}
