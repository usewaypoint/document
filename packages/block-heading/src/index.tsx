import React, { CSSProperties } from 'react';
import { z } from 'zod';

function zColor() {
  return z.string().regex(/^#[0-9a-fA-F]{6}$/);
}

const FONT_FAMILY_NAMES = [
  'MODERN_SANS',
  'BOOK_SANS',
  'ORGANIC_SANS',
  'GEOMETRIC_SANS',
  'HEAVY_SANS',
  'ROUNDED_SANS',
  'MODERN_SERIF',
  'BOOK_SERIF',
  'MONOSPACE',
] as const;

export const HeadingPropsSchema = z.object({
  props: z.object({
    text: z.string().default('Hello world'),
    level: z.enum(['h1', 'h2', 'h3']).default('h2'),
  }),
  style: z.object({
    color: zColor().optional().nullish().default(null),
    backgroundColor: zColor().optional().nullish().default(null),
    fontFamily: z.enum(FONT_FAMILY_NAMES).optional().nullish().default(null),
    fontWeight: z.enum(['bold', 'normal']).optional().nullish().default('bold'),
    textAlign: z.enum(['left', 'center', 'right']).optional().nullish().default(null),
    padding: z
      .object({
        top: z.number(),
        bottom: z.number(),
        right: z.number(),
        left: z.number(),
      })
      .optional()
      .nullish()
      .default(null),
  }),
});

export type HeadingProps = z.infer<typeof HeadingPropsSchema>;

export function Heading({
  props: { text, level },
  style: { color, backgroundColor, fontFamily, fontWeight, textAlign, padding },
}: HeadingProps) {
  const style: CSSProperties = {
    color: color ?? undefined,
    backgroundColor: backgroundColor ?? undefined,
    fontWeight: fontWeight ?? undefined,
    textAlign: textAlign ?? undefined,
    margin: 0,
    fontFamily: getFontFamily(fontFamily),
    fontSize: getFontSize(level),
    padding: getPadding(padding),
  };
  switch (level) {
    case 'h1':
      return <h1 style={style}>{text}</h1>;
    case 'h2':
      return <h2 style={style}>{text}</h2>;
    case 'h3':
      return <h3 style={style}>{text}</h3>;
  }
}

function getFontSize(level: 'h1' | 'h2' | 'h3') {
  switch (level) {
    case 'h1':
      return 32;
    case 'h2':
      return 24;
    case 'h3':
      return 20;
  }
}

function getPadding(value: HeadingProps['style']['padding']) {
  if (!value) {
    return undefined;
  }
  return `${value.top}px ${value.right}px ${value.bottom}px ${value.left}px`;
}

function getFontFamily(value: HeadingProps['style']['fontFamily']) {
  switch (value) {
    case 'MODERN_SANS':
      return '"Helvetica Neue", "Arial Nova", "Nimbus Sans", Arial, sans-serif';
    case 'BOOK_SANS':
      return 'Optima, Candara, "Noto Sans", source-sans-pro, sans-serif';
    case 'ORGANIC_SANS':
      return 'Seravek, "Gill Sans Nova", Ubuntu, Calibri, "DejaVu Sans", source-sans-pro, sans-serif';
    case 'GEOMETRIC_SANS':
      return 'Avenir, "Avenir Next LT Pro", Montserrat, Corbel, "URW Gothic", source-sans-pro, sans-serif';
    case 'HEAVY_SANS':
      return 'Bahnschrift, "DIN Alternate", "Franklin Gothic Medium", "Nimbus Sans Narrow", sans-serif-condensed, sans-serif';
    case 'ROUNDED_SANS':
      return 'ui-rounded, "Hiragino Maru Gothic ProN", Quicksand, Comfortaa, Manjari, "Arial Rounded MT Bold", Calibri, source-sans-pro, sans-serif';
    case 'MODERN_SERIF':
      return 'Charter, "Bitstream Charter", "Sitka Text", Cambria, serif';
    case 'BOOK_SERIF':
      return '"Iowan Old Style", "Palatino Linotype", "URW Palladio L", P052, serif';
    case 'MONOSPACE':
      return '"Nimbus Mono PS", "Courier New", "Cutive Mono", monospace';
  }
  return undefined;
}
