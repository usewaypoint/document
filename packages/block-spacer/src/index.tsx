import React, { CSSProperties } from 'react';
import { z } from 'zod';

export const SpacerPropsSchema = z.object({
  props: z
    .object({
      height: z.number().optional().nullish(),
    })
    .default({}),
});

export type SpacerProps = z.infer<typeof SpacerPropsSchema>;

export function Spacer({ props: { height } }: SpacerProps) {
  const style: CSSProperties = {
    height: getHeight(height),
  };
  return <div style={style} />;
}

function getHeight(height: number | null | undefined) {
  if (typeof height !== 'number') {
    return 16;
  }
  return height >= 0 ? height : 16;
}
