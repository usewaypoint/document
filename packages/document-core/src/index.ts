export { default as buildBlockComponent } from './builders/buildBlockComponent';
export { default as buildBlockConfigurationSchema } from './builders/buildBlockConfigurationSchema';
export { default as buildBlockConfigurationDictionary } from './builders/buildBlockConfigurationDictionary';

export { BlockConfiguration, DocumentBlocksDictionary } from './utils';
