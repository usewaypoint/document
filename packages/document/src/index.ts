export { default as buildBlockComponent } from './builders/buildBlockComponent';
export { default as buildBlockConfigurationSchema } from './builders/buildBlockConfigurationSchema';
export { default as buildDocumentHandlers } from './builders/buildDocumentHandlers';
export { BlockConfiguration, DocumentBlocksDictionary } from './utils';
