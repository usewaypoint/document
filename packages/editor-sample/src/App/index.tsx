import React from 'react';

import { Stack, useTheme } from '@mui/material';

import { useEditorState } from '../documents/editor/EditorContext';

import InspectorDrawer, { INSPECTOR_DRAWER_WIDTH } from './InspectorDrawer';
import SamplesDrawer, { SAMPLES_DRAWER_WIDTH } from './SamplesDrawer';
import TemplatePanel from './TemplatePanel';

export default function App() {
  const theme = useTheme();
  const [{ inspectorDrawerOpen, samplesDrawerOpen }] = useEditorState();

  return (
    <>
      <InspectorDrawer />
      <SamplesDrawer />

      <Stack
        sx={{
          marginRight: inspectorDrawerOpen ? `${INSPECTOR_DRAWER_WIDTH}px` : 0,
          marginLeft: samplesDrawerOpen ? `${SAMPLES_DRAWER_WIDTH}px` : 0,

          transition: [
            theme.transitions.create('margin-left', {
              easing: !samplesDrawerOpen ? theme.transitions.easing.sharp : theme.transitions.easing.easeOut,
              duration: !samplesDrawerOpen
                ? theme.transitions.duration.leavingScreen
                : theme.transitions.duration.enteringScreen,
            }),
            theme.transitions.create('margin-right', {
              easing: !inspectorDrawerOpen ? theme.transitions.easing.sharp : theme.transitions.easing.easeOut,
              duration: !inspectorDrawerOpen
                ? theme.transitions.duration.leavingScreen
                : theme.transitions.duration.enteringScreen,
            }),
          ].join(', '),
        }}
      >
        <TemplatePanel />
      </Stack>
    </>
  );
}
