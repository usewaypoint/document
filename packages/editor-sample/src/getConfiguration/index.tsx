import ONE_TIME_PASSCODE from './sample/one-time-passcode';
import ORDER_ECOMMERCE from './sample/order-ecommerce';
import POST_METRICS_REPORT from './sample/post-metrics-report';
import RESERVATION_REMINDER from './sample/reservation-reminder';
import RESET_PASSWORD from './sample/reset-password';
import RESPOND_TO_MESSAGE from './sample/respond-to-message';
import SUBSCRIPTION_RECEIPT from './sample/subscription-receipt';

export default function getConfiguration() {
  const hash = window.location.hash;
  if (hash.startsWith('#sample/')) {
    const sampleName = hash.replace('#sample/', '');
    switch (sampleName) {
      case 'one-time-password':
        return ONE_TIME_PASSCODE;
      case 'order-ecomerce':
        return ORDER_ECOMMERCE;
      case 'post-metrics-report':
        return POST_METRICS_REPORT;
      case 'reservation-reminder':
        return RESERVATION_REMINDER;
      case 'reset-password':
        return RESET_PASSWORD;
      case 'respond-to-message':
        return RESPOND_TO_MESSAGE;
      case 'subscription-receipt':
        return SUBSCRIPTION_RECEIPT;
    }

    console.error('Invalid sample configuration');
  }

  if (hash.startsWith('#')) {
    const encodedString = window.location.hash.slice(1);
    const configurationString = atob(encodedString);
    try {
      return JSON.parse(configurationString);
    } catch {
      console.error(`Couldn't load configuration from hash.`);
    }
  }

  const configurationString = localStorage.getItem('configuration');
  if (typeof configurationString === 'string') {
    try {
      return JSON.parse(configurationString);
    } catch {
      console.error(`Couldn't load configuration from localStorage.`);
    }
  }

  return {
    root: {
      type: 'EmailLayout',
      data: {
        backdropColor: '#e5e7e5',
        canvasColor: '#FFFFFF',
        textColor: '#242424',
        accentColor: '#0b5499',
        fontFamily: 'MODERN_SANS',
        childrenIds: [],
      },
    },
  };
}
