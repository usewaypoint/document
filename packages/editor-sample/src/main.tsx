import React from 'react';
import ReactDOM from 'react-dom/client';

import CssBaseline from '@mui/material/CssBaseline';
import { ThemeProvider } from '@mui/material/styles';

import App from './App';
import { EditorConfigurationSchema } from './documents/editor/core';
import { EditorProvider } from './documents/editor/EditorContext';
import getConfiguration from './getConfiguration';
import theme from './theme';

const DEFAULT_VALUE = EditorConfigurationSchema.parse(getConfiguration());

ReactDOM.createRoot(document.getElementById('root')!).render(
  <React.StrictMode>
    <ThemeProvider theme={theme}>
      <CssBaseline />
      <EditorProvider defaultValue={DEFAULT_VALUE}>
        <App />
      </EditorProvider>
    </ThemeProvider>
  </React.StrictMode>
);
