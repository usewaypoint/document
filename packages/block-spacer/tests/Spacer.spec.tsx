import React from 'react';

import { render } from '@testing-library/react';

import { Spacer } from '../src';

describe('Spacer', () => {
  it('renders with default values', () => {
    expect(render(<Spacer props={{}} />).asFragment()).toMatchSnapshot();
  });

  it('renders with props', () => {
    expect(render(<Spacer props={{ height: 10 }} />).asFragment()).toMatchSnapshot();
  });
});
