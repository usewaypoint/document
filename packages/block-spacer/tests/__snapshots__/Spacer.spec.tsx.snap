// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spacer renders with default values 1`] = `
<DocumentFragment>
  <div
    style="height: 16px;"
  />
</DocumentFragment>
`;

exports[`Spacer renders with props 1`] = `
<DocumentFragment>
  <div
    style="height: 10px;"
  />
</DocumentFragment>
`;
